project_redss/apply_manual_codes.py

```diff
@@ -1,127 +1,132 @@
-                    "region_coded": td["district_review"],
-                    "state_coded": td["district_review"],
-                    "zone_coded": td["district_review"],
-                    "district_coda": td["district_review"]
-                }, Metadata(user, Metadata.get_call_location(), time.time()))
+            # Up to 1 location code should have been assigned in Coda. Search for that code,
+            # ensuring that only 1 has been assigned or, if multiple have been assigned, that they are non-conflicting
+            # control codes
+            location_code = None
+
+            for plan in DatasetSpecification.LOCATION_CODING_PLANS:
+                coda_coda = plan.code_scheme.get_code_with_id(td[plan.coded_field]["CodeID"])
+                if location_code is not None:
+                    assert coda_coda.code_id == location_code.code_id or coda_coda.control_code == Codes.NOT_REVIEWED
+                if coda_coda.control_code != Codes.NOT_REVIEWED:
+                    location_code = coda_coda
+
+            # If no code was found, then this location is still not reviewed.
+            # Synthesise a NOT_REVIEWED code accordingly.
+            if location_code is None:
+                location_code = Code()
+                location_code.code_type = "Control"
+                location_code.control_code = Codes.NOT_REVIEWED
+
+            # If a control code was found, set all other location keys to that control code,
+            # otherwise convert the provided location to the other locations in the hierarchy.
+            if location_code.code_type == "Control":
+                for plan in DatasetSpecification.LOCATION_CODING_PLANS:
+                    td.append_data({
+                        plan.coded_field: CleaningUtils.make_label(
+                            plan.code_scheme,
+                            plan.code_scheme.get_code_with_control_code(location_code.control_code),
+                            Metadata.get_call_location()
+                        ).to_dict()
+                    }, Metadata(user, Metadata.get_call_location(), time.time()))
             else:
-                td.append_data({
-                    "district_coded": SomaliaLocations.district_for_location_code(td["district_coded"]),
-                    "region_coded": SomaliaLocations.region_for_location_code(td["district_coded"]),
-                    "state_coded": SomaliaLocations.state_for_location_code(td["district_coded"]),
-                    "zone_coded": SomaliaLocations.zone_for_location_code(td["district_coded"]),
-                    "district_coda": Codes.TRUE_MISSING if td["district_review"] == Codes.TRUE_MISSING else td[
-                        "district_coded"]
-                }, Metadata(user, Metadata.get_call_location(), time.time()))
+                location = location_code.match_values[0]
+
+                def make_location_code(scheme, clean_value):
```

Comment:
Is there a reason that this is defined so deeply in the code? Could we hoist it further out?